src/lexer/tokenizer.ts

```diff
@@ -0,0 +1,235 @@
+      [TokenType.NUMBER]: {
+        match:
+          /(?:0x[0-9a-fA-F]+|0b[01]+|(?:-\s*)?[0-9]+(?:\.[0-9]*)?(?:[eE][-+]?[0-9]+(?:\.[0-9]+)?)?)/,
+      },
+      [TokenType.RESERVED_CASE_START]: {
+        match: /[Cc][Aa][Ss][Ee]\b/u,
+        value: v => v.toUpperCase(),
+      },
+      [TokenType.RESERVED_CASE_END]: { match: /[Ee][Nn][Dd]\b/u, value: v => v.toUpperCase() },
+      [TokenType.RESERVED_COMMAND]: {
+        match: regex.reservedWord(cfg.reservedCommands, cfg.identChars),
+        value: v => v.toUpperCase(),
+      },
+      [TokenType.RESERVED_BINARY_COMMAND]: {
+        match: regex.reservedWord(cfg.reservedBinaryCommands, cfg.identChars),
+        value: v => v.toUpperCase(),
+      },
+      [TokenType.RESERVED_DEPENDENT_CLAUSE]: {
+        match: regex.reservedWord(cfg.reservedDependentClauses, cfg.identChars),
+        value: v => v.toUpperCase(),
+      },
+      [TokenType.RESERVED_KEYWORD]: {
+        match: regex.reservedWord(cfg.reservedKeywords, cfg.identChars),
+        value: v => v.toUpperCase(),
+      },
+      [TokenType.RESERVED_LOGICAL_OPERATOR]: {
+        match: regex.reservedWord(cfg.reservedLogicalOperators ?? ['AND', 'OR'], cfg.identChars),
+        value: v => v.toUpperCase(),
+      },
+      [TokenType.RESERVED_JOIN_CONDITION]: {
+        match: regex.reservedWord(cfg.reservedJoinConditions ?? ['ON', 'USING'], cfg.identChars),
+        value: v => v.toUpperCase(),
+      },
+      [TokenType.NAMED_PARAMETER]: {
+        match: regex.parameter(
+          cfg.namedParamTypes ?? [],
+          regex.identifierPattern(cfg.paramChars || cfg.identChars)
+        ),
+        value: v => v.slice(1),
+      },
+      [TokenType.QUOTED_PARAMETER]: {
+        match: regex.parameter(cfg.quotedParamTypes ?? [], regex.stringPattern(cfg.identTypes)),
+        value: v =>
+          (({ key, quoteChar }) =>
+            key.replace(new RegExp(escapeRegExp('\\' + quoteChar), 'gu'), quoteChar))({
+            key: v.slice(2, -1),
+            quoteChar: v.slice(-1),
+          }),
+      },
+      [TokenType.INDEXED_PARAMETER]: {
+        match: regex.parameter(cfg.numberedParamTypes ?? [], '[0-9]+'),
+        value: v => v.slice(1),
+      },
+      [TokenType.POSITIONAL_PARAMETER]: {
+        match: cfg.positionalParams ? /[?]/ : undefined,
+      },
+      [TokenType.VARIABLE]: {
+        match: cfg.variableTypes ? regex.variable(cfg.variableTypes) : NULL_REGEX,
+      },
+      [TokenType.OPERATOR]: {
+        match: regex.operator('+-/*%&|^><=.;[]{}`:$@', [
+          '<>',
+          '<=',
+          '>=',
+          '!=',
+          ...(cfg.operators ?? []),
+        ]),
+      },
+      [TokenType.STRING]: { match: regex.string(cfg.stringTypes) },
+      [TokenType.IDENTIFIER]: {
+        match: regex.identifier(cfg.identChars),
+        // type: moo.keywords({ [TokenType.RESERVED_COMMAND]: cfg.reservedCommands }), // case sensitivity currently broken, see moo#122
```

Comment:
What exactly is broken about case sensitivity here?

Reply:
actually a non-issue, I should remove this comment
moo has a keyword feature, where certain strings matched by a regex pattern can be marked as separate tokens - the example uses the strings 'if', 'else', etc.
but this can also be achieved by just having the keyword token type as its own pattern, which we are doing

we just can't use the keyword feature since it accepts a list/object of strings to catch and it would need all possible combinations of lower/upper case for each keyword - ie 'If', 'if', 'iF', 'IF'